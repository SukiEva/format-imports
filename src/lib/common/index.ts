import log4js from 'log4js';

import findUp from '@dozerg/find-up';

export function logger(category?: string) {
  return log4js.getLogger(category);
}

/**
 * Infer the function type of a class constructor.
 *
 * @example
 * ```
 *     class A {
 *       constructor(a: numbber, b?: string) {}
 *     }
 *
 *     type F = CtorToFun<typeof A>; // expect (a: numbber, b?: string) => A. Please note 'new' is removed.
 * ```
 */
type CtorToFun<T extends { new (...args: any[]): unknown }> = T extends {
  new (...args: infer A): infer B;
}
  ? (...args: A) => B
  : unknown;

/**
 * Return a function to construct a class object without 'new'.
 *
 * @example
 * ```
 *     class A {
 *       constructor(a: numbber, b: string) {}
 *     }
 *
 *     const AA = noNew(A);
 *     const a = AA(1, 'abc'); // same as new A(1, 'abc');
 * ```
 */
export function noNew<T extends { new (...args: any[]): unknown }>(c: T) {
  return new Proxy(c, {
    apply(target, args, argumentsList) {
      return new target(...argumentsList);
    },
  }) as CtorToFun<T>;
}

/**
 * Load a module from user workspace. If not found, return the default module provided.
 */
export function requireModule(moduleName: string, fromPath: string, defaultModule?: any) {
  const log = logger('format-imports.requireModule');
  const userModule = requireUserModule(moduleName, fromPath);
  if (userModule !== undefined) return userModule;
  log.warn('Cannot find', moduleName, 'from path:', fromPath, 'thus use pre-packed');
  return defaultModule;
}

declare const __webpack_require__: typeof require;
declare const __non_webpack_require__: typeof require;
const req = typeof __webpack_require__ === 'function' ? __non_webpack_require__ : require;

function requireUserModule(moduleName: string, fromPath: string) {
  const log = logger('format-imports.requireUserModule');
  const [modulePath] = findUp.sync(`node_modules/${moduleName}`, {
    cwd: fromPath,
    stopAtLimit: 1,
    type: 'directory',
  });
  if (!modulePath) return undefined;
  log.debug('Found', moduleName, 'in', modulePath);
  return req(modulePath);
}
