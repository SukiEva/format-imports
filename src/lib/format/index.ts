export function formatSource(fileName: string, sourceText: string) {
  if (fileName === 'abc') return fileName + 's';
  return sourceText;
}
