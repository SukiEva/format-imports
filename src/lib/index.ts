/* eslint-disable @typescript-eslint/no-var-requires */

export * from './config';

export * from './format/main';

/**
 * API version.
 */
export const VERSION: string = require('../../package.json').version;
