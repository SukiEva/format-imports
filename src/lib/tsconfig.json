{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "../../out/lib",
    "declaration": true
  }
}
