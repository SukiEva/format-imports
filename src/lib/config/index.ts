import minimatch from 'minimatch';

import { loadESLintConfig } from './eslint';
import { loadImportSorterConfig } from './importSorter';
import { loadTsConfig } from './tsconfig';
import { Configuration } from './types';

export {
  CompareRule,
  Configuration,
  FlagSymbol,
  GroupRule,
  KeepUnusedRule,
  SegSymbol,
  SortRules,
} from './types';

export { mergeConfig } from './helper';
export { ESLintConfig } from './eslint';

// TODO: Tests.
export function resolveConfigForFile(fileName: string, config: Configuration = {}) {
  const extConfig = loadImportSorterConfig(config, fileName);
  const eslintConfig = loadESLintConfig(fileName);
  const tsCompilerOptions = loadTsConfig(fileName);
  return { config: extConfig, tsCompilerOptions, eslintConfig };
}

// TODO: Tests.
export function isFileExcludedByConfig(fileName: string, config: Configuration) {
  const { exclude, excludeGlob, force } = config;
  if (force) return false;
  // glob
  for (const p of excludeGlob ?? []) if (minimatch(fileName, p, { matchBase: true })) return true;
  // regex
  const normalized = fileName.replace(/\\/g, '/');
  for (const p of exclude ?? []) {
    const r = new RegExp(p);
    if (r.test(fileName) || r.test(normalized)) return true;
  }
  return false;
}
