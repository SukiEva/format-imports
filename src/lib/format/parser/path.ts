import isBuiltinModule from 'is-builtin-module';
import path from 'path';

import { assertTrue } from '@dozerg/condition';

import { Configuration } from '../../config';

const NODE_PROTOCOL = 'node:';

export function normalizePath(path: string, options?: Configuration) {
  const { removeLastSlashInPath, removeLastIndexInPath, nodeProtocol } = options ?? {};
  const npFn =
    nodeProtocol === 'always'
      ? addNodeProtocol
      : nodeProtocol === 'none'
      ? removeNodeProtocol
      : undefined;
  return apply(
    path,
    normalize,
    [removeLastIndex, removeLastIndexInPath],
    [removeLastSlash, removeLastSlashInPath],
    [npFn, isBuiltinModule(path)],
  );
}

// Exported for testing purpose.
export function normalize(str: string) {
  if (!str) return '';
  const r = path.normalize(str.replace(/\\/g, '/')).replace(/\\/g, '/');
  return !str.startsWith('.')
    ? r
    : !r.startsWith('.')
    ? './' + r
    : r === '.'
    ? './'
    : r.endsWith('..')
    ? r + '/'
    : r;
}

const PATTERNS = ['index', 'index.js', 'index.jsx', 'index.ts', 'index.tsx'];

// Exported for testing purpose.
export function removeLastSlash(str: string) {
  if (!str || str === '/' || !str.endsWith('/')) return str;
  /**
   * ```txt
   * "a/"       =>  "a"
   * "index/"   =>  "index"
   * "./a/"     =>  "./a"
   *
   * "./index/" =>  "./index/"
   * ```
   */
  const parts = str.split('/');
  assertTrue(parts.length > 1);
  const prev = parts[parts.length - 2];
  if (parts.length > 2 && PATTERNS.includes(prev)) return str;
  return str.substr(0, str.length - 1);
}

// Exported for testing purpose.
export function removeLastIndex(str: string) {
  if (!str) return str;
  const parts = str.split('/');
  if (parts.length < 2) return str;
  if (!PATTERNS.includes(parts[parts.length - 1])) return str;
  /**
   * ```txt
   * "./index"        => "./"
   * "../index"       =>  "../"
   * "./index/index"  =>  "./index/"
   *
   * "index/index"    => "index"
   * ```
   *
   */
  const prev = parts[parts.length - 2];
  if (prev !== '.' && prev !== '..' && (!PATTERNS.includes(prev) || parts.length < 3)) parts.pop();
  else parts[parts.length - 1] = '';
  return parts.join('/') || '/';
}

export function removeNodeProtocol(str: string) {
  return str.startsWith(NODE_PROTOCOL) ? str.slice(NODE_PROTOCOL.length) : str;
}

function addNodeProtocol(str: string) {
  return str.startsWith(NODE_PROTOCOL) ? str : NODE_PROTOCOL + str;
}

type Fn = ((s: string) => string) | undefined;

function apply(str: string, ...norms: (Fn | [Fn, boolean | undefined])[]) {
  return norms.reduce((r, n) => (Array.isArray(n) ? (n[1] ? fn(n[0], r) : r) : fn(n, r)), str);
}

function fn(f: Fn, s: string) {
  return f ? f(s) : s;
}
