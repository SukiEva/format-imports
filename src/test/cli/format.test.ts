import {
  spawnSync,
  SpawnSyncOptions,
} from 'child_process';
import { compareSync } from 'dir-compare';
import fs from 'fs-extra';
import path, { sep } from 'path';
import tmp from 'tmp';

import { assertNonNull } from '../../lib/common';

const CMD = 'cmd.txt';
const STDIN = 'stdin.txt';
const IN_DIR = '__in';
const OUT_DIR = '__out';
const TMP_PREFIX = 'format-imports';

describe('cli/format-imports', () => {
  // change 'example' to run specific test cases.
  const examples = path.resolve(__dirname, 'examples');
  runTestSuite(examples);
});

function runTestSuite(resolved: string, relative?: string | string[]): void {
  const [name, ...rest] = relative
    ? typeof relative === 'string'
      ? relative.split(sep)
      : relative
    : [];
  if (name) return describe(name, () => runTestSuite(path.resolve(resolved, name), rest));
  const entries = fs.readdirSync(resolved, { withFileTypes: true });
  if (entries.some(e => e.name === CMD)) runTestCase(resolved);
  entries
    .filter(e => e.isDirectory() && e.name !== IN_DIR && e.name != OUT_DIR)
    .forEach(e => describe(e.name, () => runTestSuite(path.resolve(resolved, e.name))));
}

function runTestCase(resolved: string) {
  const cmds = getCmd(resolved);
  assertNonNull(cmds, 'Missing cmd.txt.');
  for (const cmd of cmds) test(`[${cmd}]`, () => runCmd(cmd, resolved));
}

function runCmd(options: string, resolved: string) {
  const stdin = getStdin(resolved);
  // create a tmp directory as the base directory to sandbox the child process.
  const tmpDir1 = tmp.dirSync({ prefix: TMP_PREFIX, unsafeCleanup: true });
  const baseDir = tmpDir1.name;
  // copy files needed to the child base directory.
  const { inDir, outDir } = getDirs(resolved);
  if (inDir) fs.copySync(inDir, baseDir);
  run(options, { stdin, baseDir });
  // setup the expected base directory
  const tmpDir2 = tmp.dirSync({ prefix: TMP_PREFIX, unsafeCleanup: true });
  const baseDirExpected = tmpDir2.name;
  if (inDir) fs.copySync(inDir, baseDirExpected);
  if (outDir) fs.copySync(outDir, baseDirExpected);
  // check base directory content.
  const r = compareSync(baseDirExpected, baseDir, { compareContent: true });
  tmpDir1.removeCallback();
  tmpDir2.removeCallback();
  expect(r).toMatchObject({ same: true });
}

function run(options: string, env?: { stdin?: string; baseDir: string }) {
  const useTsNode = process.env.USE_TS_NODE === '1' || process.env.USE_TS_NODE === 'true';
  // setup args
  const script = path.resolve(useTsNode ? 'src/bin/format-cli.ts' : 'dist/bin/format-cli.js');
  const args = [script, ...options.split(' ')].filter(a => !!a);
  // setup CWD and STDIN for child process if needed.
  const cwd = env?.baseDir;
  const stdio = env?.stdin ? [fs.openSync(env.stdin, 'r')] : undefined;
  const opt: SpawnSyncOptions = { cwd, stdio };
  const { stdout, stderr, status } = useTsNode
    ? spawnSync('ts-node-script', ['-T', ...args], opt)
    : spawnSync('node', args, opt);
  // check execution results
  expect({
    stdout: stdout.toString(),
    stderr: stderr.toString(),
    status,
  }).toMatchSnapshot();
}

function getCmd(dir: string) {
  return readFile(dir, CMD)?.split('\n');
}

function getStdin(dir: string) {
  const out = path.resolve(dir, STDIN);
  return fs.existsSync(out) && fs.statSync(out).isFile() ? out : undefined;
}

function getDirs(dir: string) {
  const inDir = checkDir(dir, IN_DIR);
  const outDir = checkDir(dir, OUT_DIR);
  return { inDir, outDir };
}

function checkDir(dir: string, name: string) {
  const r = path.resolve(dir, name);
  return fs.existsSync(r) && fs.statSync(r).isDirectory() ? r : undefined;
}

function readFile(dir: string, fn: string) {
  const file = path.resolve(dir, fn);
  if (!fs.existsSync(file)) return undefined;
  return fs.readFileSync(file).toString();
}
