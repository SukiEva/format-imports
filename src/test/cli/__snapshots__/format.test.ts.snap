// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli/format-imports directory [a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports directory config exclude [--config c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports directory config exclude [-c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports directory config exclude dry-run [--dry-run -c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
'a/b/2.ts' is excluded by config.
",
}
`;

exports[`cli/format-imports directory config exclude dry-run [-d -c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
'a/b/2.ts' is excluded by config.
",
}
`;

exports[`cli/format-imports directory config exclude force [--force -c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports directory config exclude force [-f -c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports directory config fileconfig [--config c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports directory config fileconfig [-c c/import-sorter.json a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports directory dry-run [--dry-run a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
'a/b/2.ts' will be modified.
",
}
`;

exports[`cli/format-imports directory dry-run [-d a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
'a/b/2.ts' will be modified.
",
}
`;

exports[`cli/format-imports directory no-recursive [--no-recursive a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports directory no-recursive dry-run [--dry-run --no-recursive a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
",
}
`;

exports[`cli/format-imports directory no-recursive dry-run [-d --no-recursive a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/1.ts' will be modified.
",
}
`;

exports[`cli/format-imports directory nonexist [a] 1`] = `
Object {
  "status": 1,
  "stderr": "ENOENT: no such file or directory, stat 'a'
",
  "stdout": "",
}
`;

exports[`cli/format-imports directory output [--output d a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files created.
",
}
`;

exports[`cli/format-imports directory output [-o d a] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files created.
",
}
`;

exports[`cli/format-imports file [1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config [--config a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config [-c a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config exclude [--config a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'1.ts' is excluded by config.
No files modified.
",
}
`;

exports[`cli/format-imports file config exclude [-c a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'1.ts' is excluded by config.
No files modified.
",
}
`;

exports[`cli/format-imports file config exclude force [--force -c a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config exclude force [-f -c a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config fileconfig [--config a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file config fileconfig [-c a/import-sorter.json 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file dry-run [--dry-run 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file dry-run [-d 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file nonexist [1.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "ENOENT: no such file or directory, stat '1.ts'
",
  "stdout": "",
}
`;

exports[`cli/format-imports file output [--output a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file created.
",
}
`;

exports[`cli/format-imports file output [-o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file created.
",
}
`;

exports[`cli/format-imports file output directory [--output a 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file created.
",
}
`;

exports[`cli/format-imports file output directory [-o a 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file created.
",
}
`;

exports[`cli/format-imports file output directory dry-run [--dry-run -o a 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file output directory dry-run [-d -o a 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file output dry-run [--dry-run -o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file output dry-run [-d -o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file output file [--output a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file output file [-o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "1 file modified.
",
}
`;

exports[`cli/format-imports file output file dry-run [--dry-run -o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports file output file dry-run [-d -o a/2.ts 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports files [1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config [--config c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config [-c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config exclude [--config c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/2.ts' is excluded by config.
1 file modified.
",
}
`;

exports[`cli/format-imports files config exclude [-c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'a/2.ts' is excluded by config.
1 file modified.
",
}
`;

exports[`cli/format-imports files config exclude force [--force -c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config exclude force [-f -c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config fileconfig [--config c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config fileconfig [-c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "2 files modified.
",
}
`;

exports[`cli/format-imports files config nonexist [--config c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "ENOENT: no such file or directory, open 'c/import-sorter.json'
",
  "stdout": "",
}
`;

exports[`cli/format-imports files config nonexist [-c c/import-sorter.json 1.ts a/2.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "ENOENT: no such file or directory, open 'c/import-sorter.json'
",
  "stdout": "",
}
`;

exports[`cli/format-imports files dry-run [--dry-run 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'1.ts' will be modified.
'a/2.ts' remains the same.
",
}
`;

exports[`cli/format-imports files dry-run [-d 1.ts a/2.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "'1.ts' will be modified.
'a/2.ts' remains the same.
",
}
`;

exports[`cli/format-imports files nonexist [1.ts a/2.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "ENOENT: no such file or directory, stat 'a/2.ts'
",
  "stdout": "",
}
`;

exports[`cli/format-imports files nonfile [1.ts a/2.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "Option: 'a/2.ts' is not a file.",
  "stdout": "",
}
`;

exports[`cli/format-imports files output [-o abc 1.ts 2.ts] 1`] = `
Object {
  "status": 1,
  "stderr": "Option output: should be empty if multiple files are provided.",
  "stdout": "",
}
`;

exports[`cli/format-imports help [--help] 1`] = `
Object {
  "status": 1,
  "stderr": "format-imports [options] [FILE ...]
    Format given file(s). If no files provided, read from STDIN.
format-imports [options] DIR
    Format supported files under given directory.

  -h, --help                 Show help.
  -v, --version              Show version number.
  -c, --config path::String  Read base config from a file, e.g. import-sorter.json. Each source file can have specific config which will be merged into the base config.
  -f, --force                Always format file(s) regardless of exclude config and file disable comments. Statement level disable comments will still work.
  -e, --extension String     File extension assumed when reading from STDIN and output file extension is unknown. - either: ts, tsx, js, or jsx - default: ts
  --no-recursive             Format directory recursively or not.
  -o, --output path::String  Specify the output file or directory.
  -d, --dry-run              Test running without changing any files.
",
  "stdout": "",
}
`;

exports[`cli/format-imports help [-h] 1`] = `
Object {
  "status": 1,
  "stderr": "format-imports [options] [FILE ...]
    Format given file(s). If no files provided, read from STDIN.
format-imports [options] DIR
    Format supported files under given directory.

  -h, --help                 Show help.
  -v, --version              Show version number.
  -c, --config path::String  Read base config from a file, e.g. import-sorter.json. Each source file can have specific config which will be merged into the base config.
  -f, --force                Always format file(s) regardless of exclude config and file disable comments. Statement level disable comments will still work.
  -e, --extension String     File extension assumed when reading from STDIN and output file extension is unknown. - either: ts, tsx, js, or jsx - default: ts
  --no-recursive             Format directory recursively or not.
  -o, --output path::String  Specify the output file or directory.
  -d, --dry-run              Test running without changing any files.
",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin [] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin config [--config import-sorter.json] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B
  from 'a';
import A
  from 'b';
import {
  C,
} from 'c';

export {
  A,
  B,
  C,
};
",
}
`;

exports[`cli/format-imports stdin config [-c import-sorter.json] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B
  from 'a';
import A
  from 'b';
import {
  C,
} from 'c';

export {
  A,
  B,
  C,
};
",
}
`;

exports[`cli/format-imports stdin dry-run [--dry-run] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin dry-run [-d] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension error [--extension xyz] 1`] = `
Object {
  "status": 1,
  "stderr": "Option extension: 'xyz' not one of ts, tsx, js, or jsx.

format-imports [options] [FILE ...]
    Format given file(s). If no files provided, read from STDIN.
format-imports [options] DIR
    Format supported files under given directory.

  -h, --help                 Show help.
  -v, --version              Show version number.
  -c, --config path::String  Read base config from a file, e.g. import-sorter.json. Each source file can have specific config which will be merged into the base config.
  -f, --force                Always format file(s) regardless of exclude config and file disable comments. Statement level disable comments will still work.
  -e, --extension String     File extension assumed when reading from STDIN and output file extension is unknown. - either: ts, tsx, js, or jsx - default: ts
  --no-recursive             Format directory recursively or not.
  -o, --output path::String  Specify the output file or directory.
  -d, --dry-run              Test running without changing any files.
",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin extension error [-e xyz] 1`] = `
Object {
  "status": 1,
  "stderr": "Option extension: 'xyz' not one of ts, tsx, js, or jsx.

format-imports [options] [FILE ...]
    Format given file(s). If no files provided, read from STDIN.
format-imports [options] DIR
    Format supported files under given directory.

  -h, --help                 Show help.
  -v, --version              Show version number.
  -c, --config path::String  Read base config from a file, e.g. import-sorter.json. Each source file can have specific config which will be merged into the base config.
  -f, --force                Always format file(s) regardless of exclude config and file disable comments. Statement level disable comments will still work.
  -e, --extension String     File extension assumed when reading from STDIN and output file extension is unknown. - either: ts, tsx, js, or jsx - default: ts
  --no-recursive             Format directory recursively or not.
  -o, --output path::String  Specify the output file or directory.
  -d, --dry-run              Test running without changing any files.
",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin extension js [--extension js] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension js [-e js] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension jsx [--extension jsx] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension jsx [-e jsx] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension ts [--extension ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension ts [-e ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension tsx [--extension tsx] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin extension tsx [-e tsx] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin output [--output 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output [-o 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output directory [--output 1] 1`] = `
Object {
  "status": 1,
  "stderr": "Option output: '1' is not a file.
",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output directory [-o 1] 1`] = `
Object {
  "status": 1,
  "stderr": "Option output: '1' is not a file.
",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output dry-run [--dry-run -o 1.ts ] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin output dry-run [-d -o 1.ts ] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin output file [--output 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output file [-o 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "",
}
`;

exports[`cli/format-imports stdin output file dry-run [--dry-run -o 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports stdin output file dry-run [-d -o 1.ts] 1`] = `
Object {
  "status": 0,
  "stderr": "",
  "stdout": "import B from 'a';
import A from 'b';
import { C } from 'c';

export { A, B, C };
",
}
`;

exports[`cli/format-imports version [--version] 1`] = `
Object {
  "status": 1,
  "stderr": "v1.0.0
",
  "stdout": "",
}
`;

exports[`cli/format-imports version [-v] 1`] = `
Object {
  "status": 1,
  "stderr": "v1.0.0
",
  "stdout": "",
}
`;
